sumpy/expansion/__init__.py

```diff
@@ -304,135 +378,159 @@ def get_coefficient_matrix(self, rscale):
-                for j, item in iteritems(result):
-                    coeff_matrix_transpose[j].append((i, item))
-
-            identifiers_so_far[identifier] = i
-
-        stored_identifiers = stored_identifiers
 
+        from pytools import ProcessLogger
+        plog = ProcessLogger(logger, "compute PDE for Taylor coefficients")
+
+        mis = self.get_full_coefficient_identifiers()
+        coeff_ident_enumerate_dict = dict((tuple(mi), i) for
+                                            (i, mi) in enumerate(mis))
+
+        pde_dict = self.get_pde().eq
+        for ident in pde_dict.keys():
+            if ident not in coeff_ident_enumerate_dict:
+                coeff_matrix = defaultdict(list)
+                reconstruct_matrix = []
+                for i in range(len(mis)):
+                    coeff_matrix[i] = [(i, 1)]
+                    reconstruct_matrix.append([])
+                return mis, coeff_matrix, reconstruct_matrix
+
+        max_mi_idx = max(coeff_ident_enumerate_dict[ident] for
+                         ident in pde_dict.keys())
+        max_mi = mis[max_mi_idx]
+        max_mi_coeff = pde_dict[max_mi]
+        max_mi_mult = -1/sym.sympify(max_mi_coeff)
+
+        def is_stored(mi):
+            """
+            A multi_index mi is not stored if mi >= max_mi
+            """
+            return any(mi[d] < max_mi[d] for d in range(self.dim))
+
+        stored_identifiers = [mi for mi in mis if is_stored(mi)]
+        stored_ident_enumerate_dict = dict((tuple(mi), i) for
+                                            (i, mi) in enumerate(stored_identifiers))
+
+        coeff_matrix_dict = defaultdict(lambda: defaultdict(lambda: 0))
+        reconstruct_matrix = []
+        for i, mi in enumerate(mis):
+            reconstruct_matrix.append([])
+            if is_stored(mi):
+                coeff_matrix_dict[i][stored_ident_enumerate_dict[mi]] = 1
+                continue
+            diff = [mi[d] - max_mi[d] for d in range(self.dim)]
+            for other_mi, coeff in iteritems(pde_dict):
+                j = coeff_ident_enumerate_dict[add_mi(other_mi, diff)]
+                if i == j:
+                    continue
+                # PDE might not have max_mi_coeff = -1, divide by -max_mi_coeff
+                # to get a relation of the form, u_zz = - u_xx - u_yy for Laplace 3D.
+                reconstruct_matrix[i].append((j, coeff*max_mi_mult))
+                # j can be a stored or a non-stored multi-index
+                # Look at the coeff_matrix of j to get the j as a linear combination
+                # of stored coefficients.
+                for dep, other_coeff in iteritems(coeff_matrix_dict[j]):
+                    coeff_matrix_dict[i][dep] += other_coeff*coeff*max_mi_mult
+
+        # coeff_matrix is a dictionary of lists. Each key in the dictionary
+        # is a row number and the values are pairs of column number and value.
```

Comment:
This is a bit of a wall of code. Add some comments to explain what the individual pieces are up to?